### PR TITLE
Use register hints to avoid unnecessary movs

### PR DESCRIPTION
**Pull request by jspam**
Includes a poor man's peephole optimization in the AssemblerFileGenerator

### CODE REVIEW

File: src/main/scala/mjis/AssemblerFileGenerator.scala [hunk 1]
```diff
@@ -101,7 +101,10 @@ class MjisAssemblerFileGenerator(input: AsmProgram, config: Config) extends Asse
           + ", successors: " + block.successors.map(_.nr).mkString(", "), indent = false)
         if (block.comment.nonEmpty) emit("# " + block.comment)
         block.phis.foreach(phi => emit(phiToString(phi)))
-        block.allInstructions.foreach(instr => emit(instrToString(instr)))
+        block.allInstructions.foreach {
```
Comment (Kha) [inline src/main/scala/mjis/AssemblerFileGenerator.scala]: I'd suggest turning this into a `flatMap` and just calling it `PeepholeOptimization` :) .

File: src/main/scala/mjis/RegisterAllocator.scala [hunk 2]
```diff
@@ -158,13 +161,18 @@ class FunctionRegisterAllocator(function: AsmFunction,
...
+        instr match {
+          case _: Call | _: CallWithReturn =>
+            // Block caller save registers
+            for (reg <- PhysicalRegisters.filter(CallerSaveRegisters)) {
+              // Assign size 8 to the register operand so that debug output will
+              // work nicely.
+              interval(RegisterOperand(reg, 8)).addRange(instrPos, instrPos)
+            }
+          case Mov(src: RegisterOperand, dest: RegisterOperand) if dest.regNr >= 0
+            && (PhysicalRegisters.contains(src.regNr) || src.regNr >= 0) =>
```
Comment (Kha) [inline src/main/scala/mjis/RegisterAllocator.scala]: A mov from RSP?

**Comment by Kha** [comment]
LGTM

File: src/main/scala/mjis/RegisterAllocator.scala (hunk 2 above)
Comment (jspam) [inline src/main/scala/mjis/RegisterAllocator.scala]: We cannot assign RSP as physical register to a virtual register anyway.

**Pull request merged**